### PR TITLE
[CALCITE-6111] Explicit cast from expression to numeric type doesn't check overflow

### PR DESCRIPTION
**Pull request by mihaibudiu**
This also fixes a few more tests which were disabled in SqlOperatorTests.
I still have to handle casts to decimal values, but hopefully this fixes the behavior of casts to integer types.
This behavior also has to be documented, so a future PR will describe in more details how Calcite evaluates casts.

### CODE REVIEW

**Comment by mihaibudiu** [comment]
The CI failure is from the Druid tests, so I guess this PR is ready for review.

**Comment by zstan** [comment]
I already realize the custom logic for such a cases, you can take only tests, check:
https://github.com/apache/ignite-3/pull/2786/files#diff-4ddeb0625d1cc7604c3e33e3d57ca209307a8f3e5b6aea01bee459ce06963870R675
https://github.com/apache/ignite-3/pull/2786/files#diff-ef8a24b20a4922cd2b5e92e466e3a12263444ec6887d0d87d8c885359033cf76R426

**Comment by mihaibudiu** [comment]
These tests are very useful. They cover many more cases than this PR handles. Here we only do casts.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (zstan) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13436]: what about: SELECT CAST(2147483648 as BIGINT) ?

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13436]: These cases should be covered by the test `testCastExactNumericLimits`, where I re-enabled some of the tests that were disabled.

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13436]: Well, this cast in particular should not fail, so I guess it's not covered by the NumericLimits test.
I have checked that this test would pass. 
Are you suggesting adding tests where I use the numeric limits for a type (INT) when casting to a different type (BIGINT)?

**Comment by mihaibudiu** [comment]
This is a bugfix PR, can someone please review it?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (zstan) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13423]: as for me - it`s hard to parse such a tests.
```suggestion
    f.checkFails(String.format("SELECT cast(%d+30 as tinyint)", Byte.MAX_VALUE),
```
and the same for all near

Comment (zstan) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13423]: what about implicit casts, is it all ok there ? or it out of scope ? 
SELECT 9223372036854775807 + 1;
SELECT 2147483647 + 1;

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13423]: This PR is only about casts. I plan a follow up PR about arithmetic, which has been broken for many years. There are many disabled tests because of this. That PR will handle the expressions you describe.

Regarding implicit casts: in a normal compiler, after validation the compiler should convert all implicit casts into explicit casts. This leaves many fewer cases to worry about for the subsequent stages. Unfortunately it doesn't look like Calcite works this way. I don't know enough about validation to promise I will fix this part.

However, implicit casts usually convert values from a narrow type to a wider type, so they should not be a cause of runtime failures.

But only writing lots of tests will ensure that all cases are properly handled.

**Comment by mihaibudiu** [comment]
@zstan I have rewritten the tests you suggested to make the intent clearer.
Thank you for the review, please let me know if you have more suggestions.
Note that this PR does not intend to handle casts to decimal types, that will be a separate PR.
It does not handle arithmetic overflow either, that will be yet another PR.

**Comment by mihaibudiu** [comment]
@zstan, I would like to merge this PR, since the fixes I plan to address depend on these being merged. 
Please let me know if you have more comments. I will interpret silence as an approval if you don't mind.

**Comment by zstan** [comment]
@mihaibudiu honestly, overall looks good but as for me this issue need additional eyes here, can you found one more reviewer on dev list ?

**Review by asolimando — COMMENTED** [review]
LGTM, few comments but nothing major

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (snuyanzin) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13707]: I tried to see what happens if I locally change expectedError to something else...
Surprisingly the tests continue passing
it seems that the reason is that this method is called https://github.com/apache/calcite/blob/0bec957071468a2e54a22519290ac101a752fcad/testkit/src/main/java/org/apache/calcite/test/SqlOperatorFixtureImpl.java#L157-L171

which means it checks for failure only if it is not runtime...
so it could be any expected message and any number under `cast` and it continues passing...

Or did I do something wrong?

Comment (asolimando) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13707]: @snuyanzin, I think it should fail as you say, there seems to be something wrong.

I also tried changing `runtime` to `false` just to see, in this way it always fails (with a cast with overflow, and cast without), I guess there is something broken in the test auxiliary methods we rely on here.

The source of the problem lies outside the present PR, but it's nonetheless a blocker for merging (if confirmed), as we can't rely on the unit tests.

Maybe @mihaibudiu has more familiarity with this part of the code, and can shed some light.

File: linq4j/src/main/java/org/apache/calcite/linq4j/tree/Primitive.java
Comment (asolimando) [inline linq4j/src/main/java/org/apache/calcite/linq4j/tree/Primitive.java:387]: Q: Is this used via reflection? My IDE suggests that the method is unused.

**Review by asolimando — CHANGES_REQUESTED** [review]
After @snuyanzin's comment I feel we need to get to the bottom of the test issue before merging, therefore marking the PR as "request changes" until clarified

File: linq4j/src/main/java/org/apache/calcite/linq4j/tree/Primitive.java
Comment (mihaibudiu) [reply linq4j/src/main/java/org/apache/calcite/linq4j/tree/Primitive.java:387]: Not quite reflection. The way Calcite evaluates constant expressions is to generate a Java class with a Java method which is then compiled, loaded and executed at runtime.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:664]: Yes, CALCITE-2539 seems to cover multiple issues, and I have fixed only one of them here. There are other disabled tests guarded by this Bug.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13707]: If you call the test from SqlOperatorTest it doesn't involve the evaluation.
You have to call the test from CalciteSqlOperatorTest. That class extends SqlOperatorTest but uses a test fixture which also evaluates the expressions. In the IDE you can do this easiest if you paste the following in CalciteSqlOperatorTest:

```
  @Test
  public void testOverflow() {
    super.testOverflow();
  }
```

**Comment by mihaibudiu** [comment]
> After @snuyanzin's comment I feel we need to get to the bottom of the test issue before merging, therefore marking the PR as "request changes" until clarified

I have made the fixes you suggested. The tests should be run from the class CalciteSqlOperatorTest to exercise the evaluator.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (asolimando) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13708]: Thanks for the refresher Mihai, I always forget about this behaviour in some of our test classes.

I have verified the tests, they correctly fail if a non-overflowing value/expression is passed, so they do their job.

**Comment by asolimando** [comment]
> > After @snuyanzin's comment I feel we need to get to the bottom of the test issue before merging, therefore marking the PR as "request changes" until clarified
> 
> I have made the fixes you suggested. The tests should be run from the class CalciteSqlOperatorTest to exercise the evaluator.

Thanks Mihai, the changes and the patch LGTM, can you please address the Sonar violation regarding the repeated occurrence of the string `".* out of range"`? (just define it once at the beginning of the test for better maintainability in case the error messages changes).

PS: please don't force-push while the review process is still on-going, it makes reviewers' life more difficult for not much gain, thanks!

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (snuyanzin) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13708]: Was able to test, thanks for clarification

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (snuyanzin) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:13718]: nit:
since there is already existing `org.apache.calcite.sql.test.SqlOperatorFixture#OUT_OF_RANGE_MESSAGE` would it make sense to reuse it here?

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: My last commit does exactly that. It removes one sonarcloud complaint, as pointed out by @asolimando. 

The other sonarcloud complaint is about a complex method which I didn't really write, but I slightly modified, so I didn't try to reduce the complexity. This is after all just a test method.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (asolimando) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:653]: From the variable name, it seems that `OUT_OF_RANGE_MESSAGE` was the correct error we expected (similarly to the checks in the `if` statement starting at line 635)?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (asolimando) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Yes, Sonar has false positives sometimes, it's fine to address only what's reasonable (I consider fixing the complexity of an existing test method as as nice-to-have and totally optional), but most of the time it provides good hints for keeping the code quality high.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:653]: Actually this variant of the test invokes something like `Integer.parseInt(string)`. So this is in fact a cast from a string to integer -- notice the single quotes added around the literal -- which gives a different error than a cast from a larger integer to an integer. The former produces a WRONG_FORMAT, the latter gives OUT_OF_RANGE. We could in principle trap the exception and rewrite the message, but it *is* a different exception in Java in the simplification code.

Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:653]: The integer-to-integer cast is tested a few lines above.

Comment (asolimando) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:653]: You are right, I missed the extra quote, sorry about that!

**Review by asolimando — APPROVED** [review]
LGTM, thank you, Mihai!

**Pull request merged**
